File: hot/剑指offer/38.字符串的排列/38.字符串的排列.go
package main

func permutation(s string) []string {
	track := make([]byte, 0, len(s))
	result := map[string]bool{}
	visit := make([]bool, len(s), len(s))
	for i := 0; i < len(s); i++ {
		visit[i] = false
	}
	var backTrack func()
	backTrack = func() {
		if len(track) == len(s){
			result[string(track)] = true
			return
		}
		for idx, _ := range s{
			if visit[idx] == true{
				continue
			}
			track = append(track, s[idx])
			visit[idx] = true
			backTrack()
			track = track[:len(track)-1]
			visit[idx] = false

		}
	}
	backTrack()
	res := make([]string, 0, len(result))
	for k := range result{
		res = append(res, k)
	}
	return res
}

func main() {
	s := "abc"
	println(permutation(s))
}

File: hot/剑指offer/38.字符串的排列/38.字符串的排列.py
from typing import List

"""
全排列，回溯dfs+结果set去重，visit[index]=True跳过已经使用过的元素
"""


class Solution:
    def permutation(self, s: str) -> List[str]:
        """
        回溯
        """
        result = set()  # 用集合去重
        track = list()
        n = len(s)
        visit = [False] * n

        def back_track(track):
            # 满足结束条件
            if n == len(track):
                val = ''.join(track)
                result.add(val)
                return
            for idx in range(n):
                # 剪枝
                if visit[idx]:
                    continue
                # 做选择
                track.append(s[idx])
                visit[idx] = True
                # 进入下一层
                back_track(track)
                # 撤销选择
                track.pop()
                visit[idx] = False

        back_track(track)
        return list(result)


if __name__ == '__main__':
    s = "aab"
    print(Solution().permutation(s))
